### ISSUE TITLE
Change default screen locker from XScreenSaver

### ISSUE THREAD

**Issue by mfc** [QubesOS/qubes-issues#1917]
_**Editor's note:** This issue has evolved beyond the original issue description. Please see the comments below for more recent information about this issue._

-----

#### Qubes OS version (e.g., `R3.1`): R3.1

---
### Expected behavior:

lock screen looks like log-in screen
### Actual behavior:

lock screen looks wacky-different. also ugly.
### Steps to reproduce the behavior:

lock screen in XFCE
### General notes:

see this Ask Ubuntu for reasons why this is a bad lockscreen: https://askubuntu.com/questions/216323/ugly-lock-screen-in-xubuntu

perhaps switch to gnome-screensaver as suggested in the Ask Ubuntu thread.

**Comment by marmarek** [comment]
I agree that xscreensaver is ugly. But it is effective. And given a number of problems with other screenlockers (#888, #963), not sure if it wise to change it here. Requires a lot of testing.
Actually it may be better to switch to something even uglier - physlock, in all desktop environments.

**Comment by gutsle** [comment]
Indeed there have been quite a few problems with the alternatives in the past
https://www.jwz.org/blog/2015/04/i-told-you-so-again/

**Comment by mfc** [comment]
is there any setting in xscreensaver to have it look like the XFCE log-in screen, rather than the default screensaver screen? having them look the same would be a step forward in UX, and the XFCE log-in screen certainly looks better.

**Comment by marmarek** [comment]
I'm not aware of any, unfortunately. There are settings for screen saver
itself, but not for password prompt.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by cfcs** [comment]
XScreensaver is a security disaster.
There is a long discussion / evaluation here:
https://github.com/QubesOS/qubes-issues/issues/963

It's my personal opinion that spending time on prettifying a lock screen is great, but the work should be done on a screenlocking solution that we actually want to use. Please have a look at the two suggestions below:
- https://github.com/the-cavalry/light-locker/
- `physlock` (terminal-only, so no fancy bells and whistles).

**Comment by mfc** [comment]
> XScreensaver is a security disaster.

confirmed 2026

**Comment by cfcs** [comment]
I'm going to be bold and suggest that the KDE lock screen, apart from issues with random crashes, has similar issues related to stuff popping up on top, especially at times when you (or an attacker) attaches/detaches VGA/HDMI/whatever cables, or rotates monitors and so on.

Physlock is the way to go.

**Comment by andrewdavidwong** [comment]
> XScreensaver is a security disaster.
There is a long discussion / evaluation here:
#963

No, that's a discussion of the KDE screen locker, which (as you know) is different. The only thing about XScreenSaver in that thread is a link to a problem with suspend. XScreenSaver is actually one of the more secure screen lockers available (which perhaps isn't saying much). At any rate, it does sound like physlock would be better.



>>    XScreensaver is a security disaster.

> confirmed: #2026

This, however, appears to be a legitimate security issue with XScreenSaver.

**Comment by cfcs** [comment]
@andrewdavidwong I kind of disagree, did you read the discussion? Anyway, the most important link in there, which I'd like to repost, is this one (by a KDE developer) who explains why the KDE screen locker is not sufficient, and why Xorg-based screen lockers are inherently insecure: http://blog.martin-graesslin.com/blog/2015/01/why-screen-lockers-on-x11-cannot-be-secure/

So, I restate my opinion about physlock and would like to hear some arguments against it. :)

**Comment by andrewdavidwong** [comment]
> I kind of disagree,

You kind of disagree with **what**?

> did you read the discussion?

Yes. I would not have said what I did if I had not.

> Anyway, the most important link in there, which I'd like to repost, is this one (by a KDE developer) who explains why the KDE screen locker is not sufficient, and why Xorg-based screen lockers are inherently insecure: http://blog.martin-graesslin.com/blog/2015/01/why-screen-lockers-on-x11-cannot-be-secure/

What does this have to do with anything I said?

> So, I restate my opinion about physlock and would like to hear some arguments against it. :)

What did I say that requires an argument against physlock or your opinion about physlock?

**Comment by cfcs** [comment]
>>> No, that's a discussion of the KDE screen locker, which (as you know) is different.

>> @andrewdavidwong I kind of disagree, did you read the discussion? 

> You kind of disagree with **what**?

That the discussion is about the KDE screen locker **[and all other Xorg-based screen lockers]** is irrelevant to discussion of XScreenSaver security. The link provided in my comment above highlights why XScreenSaver (and the KDE screen locker, and all the other ones) is an attempt to solve the problem using an inherently insecure approach.

>> So, I restate my opinion about physlock and would like to hear some arguments against it. :)

> What did I say that requires an argument against physlock or your opinion about physlock?

Nothing, except that XScreenSaver was one of the more secure projects around. That remark was meant for everyone, including @marmarek and people who can decide what goes into Qubes. I realize `physlock` is not very pretty, but perhaps it could be an optional thing? :)

PS: XScreenSaver nostalgia: http://insecure.org/sploits/xscreensaver.html ;)

**Comment by andrewdavidwong** [comment]
> That the discussion is about the KDE screen locker **[and all other Xorg-based screen lockers]** is irrelevant to discussion of XScreenSaver security. The link provided in my comment above highlights why XScreenSaver (and the KDE screen locker, and all the other ones) is an attempt to solve the problem using an inherently insecure approach.

You specifically said "XScreensaver is a security disaster," so, naturally, I thought you were referring specifically to... XScreenSaver. If your comment was supposed to be about all Xorg-based screen lockers, then of course I agree. This is a point I raised about Qubes years ago:

https://groups.google.com/d/topic/qubes-devel/G_wVSL9WtEk/discussion

> Nothing, except that XScreenSaver was one of the more secure projects around.

I said it's "one of the more screen lockers available (which perhaps isn't saying much). At any rate, it does sound like physlock would be better." As far as Xorg-based screen lockers go, I think it does better than most, but I also agree that Xorg-based screen lockers as a whole are problematic (again, this is all in the old thread linked above).

> That remark was meant for everyone, including @marmarek and people who can decide what goes into Qubes. I realize `physlock` is not very pretty, but perhaps it could be an optional thing? :)

When it comes to security-critical components, I don't think prettiness should matter much, and I don't think a significantly more secure solution should be merely option. It should be the default.

**Comment by marmarek** [comment]
What about light-locker? It spawns a second X server and switch to it for
the locking. There are some technical difficulties with running a second
X server in dom0, but I think it should be manageable.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by jpouellet** [comment]
We are in a unique position to be able to solve this once and for all in a more robust way than any other X-based system I am aware of.

We should be able to just **kill and restart the entire dom0 X session** (or s/dom0/GuiVM/ when we get there). This would not rely on your system not [accidentally killing the screen locker](https://lwn.net/Articles/104179/), or [accidentally making VTs available](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=697823).

I've had this idea for a while, but implementation is blocked on not some issue reconnecting to already-running gui-daemons that I haven't gotten around to debugging. The same thing needs to be fixed to be able to log out and log back in (which happens sometimes, like when restoring a large VM backup and the OOM killer gets triggered in dom0 and decides X looks like the best candidate... *sigh*) instead of needing to reboot your whole machine.

**Comment by jpouellet** [comment]
The important thing to note is that we could do this without losing state about the currently logged-in session, because the AppVM X sessions stay alive but can be rendered inaccessible in a fail-closed manner unlike on other systems with a single X session connected to both the display and the apps you wish to protect needs to stay alive.

**Comment by marmarek** [comment]
Killing dom0 X session is not harmless. GUI daemon tries its best to recover what's possible but not everything is possible. For example:
 - desktop assigned to windows will be lost (everything will appear on the same desktop)
 - tray icons will get undocked, which is most cases require restarting an application to get them back
 - restarting gui-daemon mean creating windows (at dom0 side) again, and window manager may decide to rearrange them

Some of those probably could be solved somehow, but in general touching anything around X11 handling is very risky. There are so many different toolkits, or even custom implementations that almost every change here breaks some application...

**Comment by cfcs** [comment]
@andrewdavidwong Cool, I think we're on the same page then!

@marmarek I think light-locker seems really promising. Someone (well, multiple independent people!) would have to investigate how well it manages in crash situations.

I think that killing the dom0 session to lock the screen is a bit over the top - you don't want that to happen in the middle of a dom0 update, or when doing any sort of maintenance. I usually keep a dom0 terminal window around at all times, I would hate if that was killed every time the screen went idle.

**Comment by jpouellet** [comment]
@cfcs If your only reason for wanting to retain a dom0 terminal is to ease disaster recovery in case of updates gone wrong (or other system-messed-up-ness), you can always just switch to a non-X VT (Ctrl+Alt+F2) and login there.

**Comment by Aktariel** [comment]
I'm sure this is buried in the issues or docs somewhere, but - at what point does Qubes envision transitioning to Wayland? (which would solve a heck of a lot of these "X11 lockscreens suck by design" problems.

**Comment by ksoona** [comment]
Can I at least set a screensaver instead of just the black screen? I can't install xscreensaver-extras or anything.

**Comment by cfcs** [comment]
@ksoona This thread is about making Qubes stop using an insecure screensaver, not about making the insecure screensaver prettier; you should probably file a new issue if that is what you want.

**Comment by h01ger** [comment]
FWIW, I just gave physlock a try and it worked great :) Is physlock available in fedora 25?

I also like its secure design, as opposed to other unsecure-by-design xscreenlockers... ;)

**Comment by TBK** [comment]
Apparently it is a hopeless cause on X11 - http://blog.martin-graesslin.com/blog/2015/01/why-screen-lockers-on-x11-cannot-be-secure/ the only way to solve it is to switch to Wayland (https://github.com/QubesOS/qubes-issues/issues/3366).

I experienced this scenario for the first time a couple of days ago:
`And yes that is the case: opening a context menu on any window prevents the screen locker from activating.` - source: the link above

Luckily it happened at home and only gone for a short while.

Since then I have conducted several test on several machines and it is the same, context menu blocks the screensaver from activating.

**Comment by cfcs** [comment]
@TBK I posted that link last year: https://github.com/QubesOS/qubes-issues/issues/1917#issuecomment-274088587

if you would care to read 10% the comments in the thread before replying you would have noticed that there is another way, `physlock`.

**Comment by TBK** [comment]
@cfcs My bad, should not have posted in the middle of the night.

**Comment by cfcs** [comment]
@TBK No worries. You were completely right that it's hopeless on X11, and that this is a security flaw that is even documented in `xscreensaver`.

This three year old comment by @nvesely has a proposed fix to this security flaw in Qubes in case you want to secure _your_ system (I don't think the Qubes upstream has made it clear by now that they don't consider screenlocking a security measure): https://github.com/QubesOS/qubes-issues/issues/963#issuecomment-107705261

**Comment by TBK** [comment]
@ Thanks, I will look into that.

Qubes OS is not the easiest project to to get a clear image of the state of things.

**Comment by nvesely** [comment]
I haven't used qubes in a while, but the scripts in [this repo](https://github.com/nvesely/qubes-physlock-systemd-scripts) were working as of 3 years ago.

**Comment by dylangerdaly** [comment]
I'm hitting errors using physlock (https://github.com/muennich/physlock/issues/44#issuecomment-297903096)

Issue looks to be related to PAM.

How are you guys using physlock? I want to stop relying on xscreensaver...

**Comment by dylangerdaly** [comment]
Rebuilding with SESSION=systemd and setting up the PAM module fixed it for me.

**Comment by cfcs** [comment]
@Aekez Using firejail on `xscreensaver` and the other X-based screensavers actually increases the likelihood that they will crash, meaning that it increases the attack surface for someone looking to bypass your lock screen. For `physlock` I don't think it would make much of a difference, but wouldn't hurt.

**Comment by Aekez** [comment]
@cfcs ah, that's unfortunate if it can make it even worse. Thanks for sharing the insight though.

**Comment by andrewdavidwong** [comment]
Increasing priority to **critical** based on discussion in #2026.

**Comment by na--** [comment]
I haven't tried physlock, but I was using light-locker until very recently, and I had to switch back to xscreensaver because of https://github.com/QubesOS/qubes-issues/issues/4943 :disappointed: The latest dom0 kernel that light-locker didn't have issues with 4.14.103 IIRC... :man_shrugging: 

I only have the most surface-level knowledge of how these things work, and I have no idea about the root cause of my issues, so I'm not sure if physlock will have some of the same problems, :crossed_fingers: that it turns out to be a secure and problem-free locker :sweat_smile: I probably can't be very useful in actually implementing this, but I'm very eager to beta-test anything... :eyes:

**Comment by martin-paulus** [comment]
@andrewdavidwong @marmarek Did you consider https://tools.suckless.org/slock/ ? It was not mentioned in this issue, but I think it fits in with the Qubes OS goals: prioritize security and privacy, and avoid complexity where possible. I especially like that it doesn't feature a username or password field so nothing can be gained by shoulder surfing (unless it's your hands that are being observed). And it's already available in the dom0 package repos.

**Comment by andrewdavidwong** [comment]
> @andrewdavidwong @marmarek Did you consider https://tools.suckless.org/slock/ ? It was not mentioned in this issue, but I think it fits in with the Qubes OS goals: prioritize security and privacy, and avoid complexity where possible. I especially like that it doesn't feature a username or password field so nothing can be gained by shoulder surfing (unless it's your hands that are being observed). And it's already available in the dom0 package repos.

I wasn't aware of it until now. I just read the page you linked, but it's very brief. Based on your description, it sounds like the lock screen would be just one solid color, and the user has to blindly type in their password to unlock. This sounds like it might not be the best UX for general users. CC: @ninavizz

**Comment by rustybird** [comment]
Also, slock is unfortunately still X11-based (unlike physlock), so it can't really solve the fundamental problems.

**Comment by 3hhh** [comment]
+1 from me for [physlock](https://github.com/muennich/physlock).

Reasons:
- < 1k code lines = auditable
- superior design decisions (does not rely on X for screen locking) = likely more secure
- relatively few dependencies (Linux kernel, PAM & systemd)
- takes care of other open logins (VTs) & SysRqs
- suspend & hibernate support
- xss-lock can be used to make it play nicely with X
	- instead of "blank screen", physlock will take over
	- xset s can be used
	- probably even GUI etc. can be used for configuration
- UX should never be superior to security (yes, it has no moving images etc. and will never get any by design, it only says: "Password:")

Cons:
- apparently requires root password to be set
- had some issues with my dock / did not unlock (but xscreensaver had worse issues with that by unlocking randomly...)

Instructions for Qubes OS so that maybe some people test it:
(Marek was fully right in 2016 to say that it requires quite some testing.)

`sudo qubes-dom0-update gcc make pam-devel systemd-devel xss-lock`
verify tag signatures
see physlock [README](https://github.com/muennich/physlock)
in particular follow PAM instructions (or run into an endless "authentication failed" loop)

set physlock as screen locker after X minutes:
via xfce4:
```
sudo su
passwd #root pw required to be set 
#make sure `/etc/xdg/autostart/xfce4-xss-lock.desktop` exists with `xss-lock xflock4` (does exist in Qubes OS 4)
xfconf-query -c xfce4-session -p /general/LockCommand -s "/usr/local/bin/physlock -ms" --create -t string
```
create a `/etc/xdg/autostart/xset.desktop` as found here (or set the timeout using another method):
```
[Desktop Entry]
Name=xset
Comment=Set screensaver timeout
Exec=bash -c 'sleep 60 && xset s 610'
Terminal=false
Type=Application
StartupNotify=false
```
If you need audio when locked:
`sudo usermod -a -G audio [your user]`


------------------------------------

An alternative which is likely better maintained, has a security focus, all the bells and whistles, is in newer Fedora repos, but suffers from the X problems nonetheless (at least they admit it):
https://github.com/google/xsecurelock#security-design

Ah yes, and some may not like google. ^^

**Comment by martin-paulus** [comment]
> andrewdavidwong : 
> Based on your description, it sounds like the lock screen would be just one solid color, and the user has to blindly type in their password to unlock. This sounds like it might not be the best UX for general users.

It has three colors: initially black, then blue while typing, and red to show you just failed authentication. Considering a general user _for Qubes OS_ using this kind of locker does not sound outlandish to me. But of course, you are entitled to your opinion.

> 3hhh 
> takes care of other open logins (VTs) & SysRqs

In light of Andrew's comment about user-friendliness, and considering the efforts the Qubes OS project has put in creating a GUI which draws attention to qubes domain boundaries, why do we bother with virtual consoles at all?

Consider disabling virtual consoles, so that only the Qubes OS GUI is available:

dom0: /etc/systemd/logind.conf
```
[Login]
- #NAutoVTs=6
- #ReserveVT=6
+ NAutoVTs=0
+ ReserveVT=0
```

This will limit the requirements of any (replacement) screenlocker to securing the GUI only.

**Comment by 3hhh** [comment]
> Instructions for Qubes OS so that maybe some people test it:
> (Marek was fully right in 2016 to say that it requires quite some testing.)
> 
> `sudo qubes-dom0-update gcc make pam-devel systemd-devel xss-lock`
> verify tag signatures
> see physlock [README](https://github.com/muennich/physlock)
> in particular follow PAM instructions (or run into an endless "authentication failed" loop)
> 
> set physlock as screen locker after X minutes:
> via xfce4:
> 
> ```
> sudo su
> passwd #root pw required to be set 
> #make sure `/etc/xdg/autostart/xfce4-xss-lock.desktop` exists with `xss-lock xflock4` (does exist in Qubes OS 4)
> xfconf-query -c xfce4-session -p /general/LockCommand -s "/usr/local/bin/physlock -ms" --create -t string
> ```
> 
> create a `/etc/xdg/autostart/xset.desktop` as found here (or set the timeout using another method):
> 
> ```
> [Desktop Entry]
> Name=xset
> Comment=Set screensaver timeout
> Exec=bash -c 'sleep 60 && xset s 610'
> Terminal=false
> Type=Application
> StartupNotify=false
> ```
> 
> If you need audio when locked:
> `sudo usermod -a -G audio [your user]`

After a while of testing I'd now recommend to use a script instead of using physlock directly:
```{.bash}
#!/bin/bash

function isRunning {
pgrep -a '^physlock$'
}

#parse args
keep_open=1
if [[ "$1" == "--keep-open" ]] ; then
	keep_open=0
	shift
fi

#NOTE: for some reason the full path is required below for xss-lock
isRunning || { /usr/local/bin/physlock -dms "$@" ; sleep 1 ; }

#Idea:
#make xss-lock think that it controls the screenlocker, but in fact it doesn't
#reason: xss-lock may crash and we don't want it to take down the screen lock
if [ $keep_open -eq 0 ] ; then
	stime=10
	while isRunning ; do
		echo "Sleeping for ${stime}s..."
		sleep $stime
	done
fi
exit 0
```

I.e. use that for `xfconf-query` and hotkeys etc.

I guess some official doc on how to deploy an alternative screen locker would be nice in total - regardless of which one becomes the next default.

**Comment by DemiMarie** [comment]
Given that #3898 is a duplicate of this one, should this be backported to R4.0 as well?

**Comment by dylangerdaly** [comment]
When trying to use physlock a few months ago, I got x11 artifact, I'll give this another shot and see if it's working better now

**Comment by ninavizz** [comment]
The most important thing with a screenlocker wrt usability, is that it not make an impression with the user that it's malware. In the context of Qubes, this is important. One of my first impressions of Qubes, was that the xscreen lock looks like malware. I even approached it's developer dude about changing that, but he finds that aesthetic to be charming so told me to eff-off (literally—yep, Jamie Zawinski is That Guy).

Second, is that there be an option for users to see their password as they type it. The current screenlocker also fails in this capacity.

I'm not a developer so can't give things a test run on a dev box. I'd love to see screenshots of what is being proposed for 4.1, and am STOKED this might go in! Yes, it is a big priority to me, and has also been flagged by others at FPF as being desirable.

**Comment by 3hhh** [comment]
> The most important thing with a screenlocker wrt usability, is that it not make an impression with the user that it's malware. In the context of Qubes, this is especially important—doubly so, if there's a desire to get non-techy users to adopt Qubes. One of my first impressions of Qubes, was that the xscreen lock looks like malware. I even approached it's developer dude about changing that, but he finds that aesthetic to be charming so told me to eff-off (literally—yep, Jamie Zawinski is That Guy).

If malware had a particular look, the job of AV vendors would be terribly easy.

> Second, is that there be an option for users to see their password as they type it. The current screenlocker also fails in this capacity.

Security-wise that's an anti-feature as it opens up simple shoulder-surfing attacks.

> I'm not a developer so can't give things a test run on a dev box. I'd love to see screenshots of what is being proposed for 4.1, and am STOKED this might go in! Yes, it is a big priority to me, and has also been flagged by others at FPF as being desirable.

I think you want something more visually appealing. `physlock` won't do that - it looks like a Ctrl-Alt-F1 tty logon.

Anyway if you want something visually appealing you'll need the X server for lovely graphics etc. As described above that's insecure. So currently there's these options:

X server screensavers: less secure, possibly more visual appeal
non-X screensavers: possibly more secure, likely less appealing

`xscreensaver` belongs to the first category with low security properties, but ironically it isn't visually appealing either; `physlock` belongs to the second category.

**Comment by DemiMarie** [comment]
Another idea I had is for the screenlocker to be the display manager itself, and to use `PRCTL_SET_DEATHSIG` so that if the screenlocker crashes for any reason, X dies with SIGKILL.

**Comment by 3hhh** [comment]
> Another idea I had is for the screenlocker to be the display manager itself, and to use `PRCTL_SET_DEATHSIG` so that if the screenlocker crashes for any reason, X dies with SIGKILL.

So a non-X-screensaver. ;-)

But yes, it's an option. X server restarts aren't too good for user experience neither though. I'm not sure whether physlock causes less X restarts though (mine dies every now and then and especially after physlock, but it may be related to awesome).

Essentially there doesn't seem to be a perfect solution which ticks all boxes wrt security, usability and visual appeal.

**Comment by DemiMarie** [comment]
Actually, there is: we can switch to Wayland.  That brings all sorts of improvements, but is also a lot of work.

**Comment by jpouellet** [comment]
On Mon, Nov 30, 2020 at 1:51 PM Nina Eleanor Alter <notifications@github.com>
wrote:

> The most important thing with a screenlocker wrt usability, is that it not
> make an impression with the user that it's malware. In the context of
> Qubes, this is important. One of my first impressions of Qubes, was that
> the xscreen lock looks like malware. I even approached it's developer dude
> about changing that, but he finds that aesthetic to be charming so told me
> to eff-off (literally—yep, Jamie Zawinski is That Guy).
>

On the contrary, I suspect that many Qubes users believe the most important
thing with a screenlocker, is that it actually, you know, *lock your
screen*, and do so in the most safe, reliable, and
least-likely-to-be-bypassed manner. After all, these properties are
directly security relevant, whereas appearance of the locker is only
indirectly so, if at all.

In the author's defense, "That Guy" happens to have laid out specific
technical security arguments (16 years ago!) against the risk which would
be posed by making xscreensaver "prettier" in the usually-proposed manners,
which can be found at https://www.jwz.org/xscreensaver/toolkits.html

And as it turns out, history happens to have shown the author's stance on
this particular matter to have been prudent and correct.

I'd be willing to guess that the tone of the author's response is nothing
personal, but comes from frustration with responding to the same
question/suggestion over and over again since initially writing
xscreensaver 28 years ago (!), and having written a detailed explanation of
why they aren't interested in adopting the proposed suggestion 16 years
ago. Anyone gets tired of repeating themselves eventually.

Now, this isn't to say that we can't have nice things (we certainly can!)
and also not to say that xscreensaver should be considered the last word in
screen locking either (one need only note the numerous observations of
notifications leaking through, or failure-to-lock situations when another
program holds XGrab{Pointer,Keyboard,...} to realize that).

There are various alternative approaches which may be considered, each with
their own trade-offs. These include (among other things):
1. Using a VT-switching+lockdown approach such as physlock.
2. Using something like https://github.com/google/xsecurelock
3. "locking the screen" in the most-unfriendly yet most-reliable manner
that Qubes is uniquely positioned to leverage: by killing the X session
entirely and "unlocking the screen" by logging in again and restarting the
gui daemons (which restore windows, positions, titles, etc. from the
still-running gui agents in the running guests).
4. The various options available in wayland, which is fundamentally better
positioned to safely solve screen locking than any X11-based screen locker
is (due to architectural deficiencies in the X11 protocol, modulo
extensions which never took off). This would of course require quite a bit
of prerequisite work.

Xscreensaver is what it is, and it is understandable for the
author/maintainer to refuse suggestions which they deem inappropriate, or
simply do not like. If we don't like the result, we are free to pick
another alternative (see above for suggestions) or fork it (hint, we
already have [1], though substantially changing xscreensaver in particular
is inadvisable -- see above-linked argument for why).

[1]: https://github.com/QubesOS/qubes-xscreensaver

I don't mean to discourage any efforts at usability. I do support and
appreciate them! I just want to raise awareness that in this case it can be
at odds with security for real yet perhaps non-obvious reasons.

IMO, questions such as "How hard is the lock screen to bypass?" should be
given primary consideration, including architectural / design
considerations of theoretical bypasses which postulate the existence of
bugs which are not presently known and cannot easily be demonstrated, but
can be reasonably anticipated, for example in the arguments laid out by
"That Guy" 16 years ago.

Confidence in the absence of edge-cases resulting in failure to guarantee a
locked screen is, I would argue, more important than how it looks when
allegedly locked.

Hopefully this won't be interpreted the wrong way. I really do bear no ill
will, I just appreciate the author's work, as well as appreciate efforts to
improve Qubes usability, and am sad to see them at odds. I hope this
additional background helps.

>

**Comment by DemiMarie** [comment]
So one idea I had is:

- Have the screenlocker run as the *parent* of the X server.
- The screenlocker uses ptrace on the X server with `PTRACE_O_EXITKILL` set.
- The screenlocker receives the “I need to lock the screen” command over IPC, such as D-Bus.
- If the screenlocker fails to lock the screen, or if an internal error occurs, it exits.
- If the screenlocker exits, the Linux kernel immediately and forcibly terminates the X server with `SIGKILL`, and the user is returned to the login screen.

The main advantage of this approach is that the only way it can fail is through a kernel vulnerability.  `PTRACE_O_EXITKILL` is intended for ptrace-based sandboxing programs, meaning that if it doesn’t work as documented, that’s a security hole in the kernel.

**Comment by ninavizz** [comment]
Pretty is not the goal. Inconsistent graphics and fatalistic metaphors (a wall on fire), are. 
In 16 years of doing user research and design, this has consistently been an observation in users.

Malware does have an archetypical aesthetic in the mind of users; anything that seems inconsistent or out of place, does... which is why consistency in execution and design is so important. Same, with fatalistic metaphors. As security experts, you all know that malware has gotten much smarter since The Matrix and Wargames instilled those metaphors onto popular culture, but their dogmas remain with users.

Likewise, the shoulder-surfing concern of security experts, is a frequent discussion among security UX folks—and user agency with access to usability, consistently wins out. By default the password should be obfuscated, but denying users the option to even peek at what they've typed, I don't feel is reasonable. It is harder to educate users and to encourage best-practices in opsec via ux, but it is a worthwhile challenge.

When security apps don't do this, users will adapt what they can—their behavior and choices—to much less secure options. In the case of passwords, things much easier to type and to subsequently hack. Human behavior is irrational. That has to be considered and as technologists, we need to adapt to that. As an example—I typically use "easy to type" 6 word diceware passphrases... which are almost impossible to get right the first few times. It's exhausting, always re-typing w/o knowing. So my workaround in Qubes, is to just use my phone number. I'm lazy and stubborn. Simply telling a user "that is unsafe" won't change their behavior. In enterprise environments where employees will be fired for sharing passwords, it's still common to see passwords on post-it notes everywhere. That's easier than using a password manager; so the rational/reasoned argument that "it's risky so therefore they won't do it when we tell them not to" doesn't play-out in reality.

At the end of the day, humans will human, and the goal with all inclusive UX needs to be reducing anxiety and improving paths to secure behavior—but still giving the user agency to make choices. If that makes sense. 

My own goal with all of the Qubes UX, is not to specifically serve today's users to the exclusion of non-technical high-risk users, but to address all of the above. Which is hard.

**Comment by na--** [comment]
FWIW, I am someone who, if given a simple choice between security and UX, would prefer security 99% of the time. But one reason I muddled with light-locker for a long while (https://github.com/QubesOS/qubes-issues/issues/1917#issuecomment-658658466) was because xscreensaver has _terrible_ UX. 

The biggest issue I have with it happens because I have 2 different keyboard layouts configured, my laptop keyboard doesn't have visual caps lock or num lock indication, and the laptop designers chose to have part of the normal keys double as the numpad (jkluio are 123456 if numlock is on :man_facepalming: ). So, given that xscreensaver doesn't have a way to show the current keyboard layout or the caps lock and num lock key state, nor a way to show what I'm typing, I have no idea in which of the 8 (2^3) possible states I am currently in, and may end up having to re-type my password several times. Or dozens of times, if I mistype it when I'm in the right state... I get why shoulder-surfing might be a problem, but re-typing your password a lot is kind of the same issue, and I'd prefer to be able to see what I type when I'm alone in a room...

I will try physlock soon, if I understand how it works correctly, that should at least remove the keyboard layout ambiguity and reduce the number of possible states to 4. And replacing my laptop in the future with something saner might fix _my_ problems altogether, I can live with entering my password in a TTY. But the holy grail of having a screenlocker that is both secure and has a nice UX would be a dream come true...

**Comment by rustybird** [comment]
> * If the screenlocker fails to lock the screen, or if an internal error occurs, it exits.
> 
> * If the screenlocker exits, the Linux kernel immediately and forcibly terminates the X server with `SIGKILL`, and the user is returned to the login screen.

That's an interesting twist for the classic scenario "I accidentally pressed the `Q` button (Applications Menu) while lowering the notebook lid, so the screensaver couldn't engage": Now it fails closed, but paranoid users are going to have a heart attack when they come back and it looks like someone tampered with their screensaver...

Also, is the X server really killed _immediately_? If an attacker is pointing a video camera at the screen and they manage to crash the screensaver, could the camera capture a still frame of the unlocked X server before it's killed?

**Comment by DemiMarie** [comment]
@rustybird “immediately” as in “cannot make any more system calls”.

**Comment by h01ger** [comment]
On Tue, Dec 01, 2020 at 10:11:56AM -0800, Jean-Philippe Ouellet wrote:
> 3. "locking the screen" in the most-unfriendly yet most-reliable manner
> that Qubes is uniquely positioned to leverage: by killing the X session
> entirely and "unlocking the screen" by logging in again and restarting the
> gui daemons (which restore windows, positions, titles, etc. from the
> still-running gui agents in the running guests).

I actually like this idea (and your whole message, Jean-Philippe), though
I have to say this does not work with i3: all the windows and their content
is restored, but the window positions are not. (I still think that's
impressive.)


-- 
cheers,
	Holger

 ⢀⣴⠾⠻⢶⣦⠀
 ⣾⠁⢠⠒⠀⣿⡁       holger@(debian|reproducible-builds|layer-acht).org
 ⢿⡄⠘⠷⠚⠋⠀ PGP fingerprint: B8BF 5413 7B09 D35C F026 FE9D 091A B856 069A AA1C
 ⠈⠳⣄

If secure encryption is outlawed, only criminals will have it.

**Comment by brendanhoar** [comment]
I like the idea as well. 

I will just note that it does come with some consequences under 4.0.x which is that dom0 windows won’t be restored, only VM windows. So, if scripting VMs (via manually invoked python/bash) you’d need to keep that in mind.

Unsure of related impacts on GUI-vm approaches in 4.1/future versions of Qubes.

B

**Comment by jpouellet** [comment]
On Tue, Dec 1, 2020 at 5:01 PM Holger Levsen <notifications@github.com> wrote:
> I actually like this idea (and your whole message, Jean-Philippe), though
> I have to say this does not work with i3: all the windows and their content
> is restored, but the window positions are not. (I still think that's
> impressive.)

For which we have https://i3wm.org/docs/layout-saving.html ! ;)

**Comment by jpouellet** [comment]
The current pain-points of that approach are actually:
1. tray icon windows (like nm-applet) which seem to get lost unless
restarted (and I'm sure could be fixed, I just haven't bothered). and
2. of course dom0 windows, which you shouldn't have many of anyway,
and I suspect may be entirely a non-issue with the GUIVM approach
where the GPU is still owned by dom0 but has only a single window in
which the GUIVM's WM/DE performs compositing.

**Comment by marmarek** [comment]
3. Window to workspace assignments get lost.

**Comment by brendanhoar** [comment]
> 1. Window to workspace assignments get lost.

That’d be an issue for a significant number of users, yes.

**Comment by jpouellet** [comment]
On Tue, Dec 1, 2020 at 5:22 PM Marek Marczykowski-Górecki
<notifications@github.com> wrote:
> Window to workspace assignments get lost.

True, I'd forgotten about this.

For i3, this is solved by the generic layout save/restore functionality.

For a general WM-independent solution, it seems one would need to
somehow serialize and restore the _NET_WM_DESKTOP window property [1]
for the windows. It's been a while since I've looked at any of this
stuff, so please correct me if I'm wrong, but, just leaving notes from
memory in case anyone wants to pursue it: I believe this serialization
would need to happen in the gui-daemon xside, because IIRC the guests
have no concept of desktops/workspaces. Unfortunately, IIRC no state
is retained at all by the gui-daemon xside code, and it is "stateless"
in the sense that it just gets told what guest it should service and
rehydrates its state from the gui-agent (vmside code). This is
unfortunate, because it means the state is not already conveniently
available in the place where it is being persisted.

As just an observation: Xfce does have a mechanism already for
saving/restoring session state, but it appears to be (currently)
primarily based on the command line of the program as it was started,
and not its runtime state. The session files are at
~/.cache/sessions/xfce4-session-dom0:0, and the code which generates
these save files is at [2] and [3]. See for example
xfsm_client_save_restart_command(), which uses ps to fetch the command
args. One might consider extending this session state with another
property which records which desktop the windows were on.

Adding WM-agnostic which-desktop serialization to the gui-agent feels
slightly "cleaner" than accumulating WM-specific approaches, but I'm
really not fond of adding a notion of persistent state to the gui
daemon at all either. Intentional lack of state is an attractive
property.

[1]: https://specifications.freedesktop.org/wm-spec/wm-spec-1.3.html#idm45315510687072
[2]: https://github.com/xfce-mirror/xfce4-session/blob/master/xfce4-session/xfsm-client.c
[3]: https://github.com/xfce-mirror/xfce4-session/blob/master/xfce4-session/xfsm-properties.c

Just some notes for anyone who might be interested in exploring further.

**Comment by jpouellet** [comment]
On Tue, Dec 1, 2020 at 7:53 AM 3hhh <notifications@github.com> wrote:
>
>> Another idea I had is for the screenlocker to be the display manager itself, and to use PRCTL_SET_DEATHSIG so that if the screenlocker crashes for any reason, X dies with SIGKILL.
>
> So a non-X-screensaver. ;-)
>
> But yes, it's an option. X server restarts aren't too good for user experience neither though. I'm not sure whether physlock causes less X restarts though (mine dies every now and then and especially after physlock, but it may be related to awesome).
>
> Essentially there doesn't seem to be a perfect solution which ticks all boxes wrt security, usability and visual appeal.

I think that conclusion misses some nuance.

Approaches such as physlock (the VT-switching+locking category) do not
inherently / necessarily need to be ugly. In short, because linux has
a rather mature and capable framebuffer subsystem independent of X or
Wayland or anything else. One can open /dev/fb0 and write to it to
very simply display arbitrary graphics (of whatever arbitrary visual
style you want) on the screen, from an otherwise "boring" / "ugly" tty
which you are locked to.

The main problems with X screen locking security is because of the
complexity and inherent design deficiencies of the APIs involved when
trying to perform screen locking within the *same* X server instance
that simultaneously has your sensitive X client programs connected to
it, as well as the locking program, as well as any programs in dom0
which can directly grab input devices, as well as being the same thing
responsible for handling I/O during the time that your screen is
locked and you might not be the one sitting in front of your computer.
The conclusion that "X has problems, therefore X can't be involved at
all" is arguably an oversimplification. A scheme with multiple X
servers, which switches away from your logged-in user X session (and
perhaps optimistically tries to lock that using a conventional
mechanism for defense in depth anyway) to a "locker-only" X session,
and inhibits VT-switching, may potentially avoid the primary class of
issues with X screen lockers. This might sound crazy, but it actually
isn't, and is exactly how some display managers work today. X crashes
with VT switching were a thing historically, but there's been a lot of
rework in the kernel graphics subsystems, and the linux graphics stack
should be much more resilient by design to that entire category of
issues now. Read up on KMS / DRM / DRI{2,3} and the corresponding X
server changes for details. Long story short: device access is now
much more heavily mediated by the kernel, which imposes constraints on
appropriate driver state transitions.

That said, even if one uses a scheme with dedicated "session" /
"locker" X servers, some security questions, like those of complexity,
device access, etc. still remain, but... X need not be involved at
all!

Modern display management in linux being controlled by the kernel and
all relevant driver state tracked fairly well these days means that
using a plain old linux framebuffer (X / Wayland strictly optional) to
get "pretty" (read: visually whatever you want) password-input prompts
is also feasible.

I'd personally still prefer the simplicity of an approach which just
reads characters from a tty, since I don't care what my screen locker
looks like and really just care that it be safely implemented (with
bonus points for assisting with anti-shoulder-surfing input
mechanisms, and possibly make available some lightweight form of
remote attestation which I may optionally elect to verify before
entering my password), but, I admit that I'm probably on the more
security-focused end of the spectrum of users.

**Comment by rustybird** [comment]
@DemiMarie:
> “immediately” as in “cannot make any more system calls”.

With just `PTRACE_O_EXITKILL`? And not `PTRACE_SYSCALL`, which I assume has too much overhead to be used on the X server.

Barring that, [`do_exit()`](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/kernel/exit.c?h=v5.9.12#n710) looks like it would close the screensaver's file descriptors (including the X client socket) via `exit_files()`, then do a bunch of other stuff, then kill the child process via `exit_notify()` -> `forget_original_parent()` -> `exit_ptrace()`. So in the meantime the X server (unless somehow pinned to the same CPU as the screensaver?) could theoretically notice the socket shutdown, destroy the screensaver window, and arrange for the windows behind it to repaint. Maybe that's unrealistic timing, or I'm missing something obvious.

**Comment by woju** [comment]
As a data point, I use i3lock, which draws uses xcb and cairo, but does not use the worst offenders (like accessibility stack): https://github.com/i3/i3lock#requirements. It's visually nicer that xscreesaver, I can set a wallpaper and the feedback to user input is a nice flashing circle in the middle.

WRT to window->screen mapping, in awesome this is mostly non-issue if you have set up the rules. The worst offenders are systray icons from nm-applet, which for some reasons become separate windows (they dock again to tray when closed).

**Comment by 3hhh** [comment]
> 3. "locking the screen" in the most-unfriendly yet most-reliable manner
> that Qubes is uniquely positioned to leverage: by killing the X session
> entirely and "unlocking the screen" by logging in again and restarting the
> gui daemons (which restore windows, positions, titles, etc. from the
> still-running gui agents in the running guests).

4. issue: Paused VMs (non-running `qubes-gui-daemon`), but that leads to funny effects without X restart atm anyway (window hanging somewhere & see #5988).

> Have the screenlocker run as the parent of the X server.
> The screenlocker uses ptrace on the X server with PTRACE_O_EXITKILL set.
> The screenlocker receives the “I need to lock the screen” command over IPC, such as D-Bus.
> If the screenlocker fails to lock the screen, or if an internal error occurs, it exits.
> If the screenlocker exits, the Linux kernel immediately and forcibly terminates the X server with SIGKILL, and the user is returned to the login screen.

That scrensaver would have to run outside of that X (i.e. framebuffer or so) or inside a dedicated instance as mentioned by @jpouellet (nice posts btw!) though to avoid the usual X display issues that don't kill X, but are harmful anyway (focus grabbing etc.).
Moreover the issues 1.-4. also apply here in the last step, but they are more of an edge case / happen more rarely.

Both approaches are generic (nothing specific to any WM or display server), which I like.
Both also indicate that looking into better Qubes OS resilience against X server restarts is likely worth the time.

**Comment by 3hhh** [comment]
Projects related to X session saving (aka "session managers"):
https://github.com/christarazi/sessionctrl
https://github.com/nlpsuge/xsession-manager
https://gitlab.freedesktop.org/xorg/app/xsm

Most seem to be in early dev or stale though.

**Comment by 3hhh** [comment]
> https://github.com/christarazi/sessionctrl
> https://github.com/nlpsuge/xsession-manager

Hmmm I checked the code a bit and it's not really usable for Qubes OS as they heuristically use the command-line (which is empty for VM windows) as identifier to restore window positions etc.

Anyway the idea to fix the aforementioned X window restore issues 2 & 3 still holds:

1. On screensaver start (using e.g. the screensaver-as-parent method), save the session.
2. On unexpected X restart, restore the session (e.g. on `/etc/xdg/autostart`), i.e. window positions & desktop assignments, and delete it afterwards. Maybe display some explanatory notification.
3. (Optional): On graceful shutdowns, delete saved sessions.

For VM windows, use the window ID of the X server inside the VM (see `xprop`: `_QUBES_VMWINDOWID` seems to hold it).
If no matching window is found, ignore it.
There is no way to restore dom0 windows unfortunately (they get terminated if X dies) unless one wants to go the "we simply re-start GUI processes based on their command-line (`ps -eo args`)" path, which I deem confusing as the GUI applications don't always continue at where they left off (e.g. terminal).
With the @DemiMarie "blue pill" approach it should only happen 1/10 times or less anyway.

At first sight it looks like `wmctrl` and `xprop` should be able to do this, but I tested `wmctrl` window moving and workspace assignments on awesome and it does pretty much nothing at all except some flickering.
So I guess the restore can unfortunately only be handled per WM. :(

Btw it would be unwise to kill X on purpose (@jpouellet option 3) as that would also terminate e.g. `qvm-backup` running in some dom0 terminal.

In total it seems to me that locking down the VTs might still be more simple unless one needs fancy stuff.
For fancy stuff one could still lock them down and start a dedicated X instance for whatever screensaver the user wishes and then, if that screensaver finished gracefully, unlock the VTs incl. the first X instance again. If the screensaver X instance crashes... blue pill and do it again? At least no windows to restore... ^^

**Comment by ninavizz** [comment]
Sooo many new comments here, and I've been slammed w/ a firedrill project for a new client for the past 2 weeks, so I'll just post here to contribute the below:

Someone up above cited "Picking security over user experience" as a necessity. As a UX practicioner who specializes in security, I gotta push back on that. There are a small handful of us, who care (deeply!) about both—and who've dedicated our work in recent years, to always solving both problems, together. 

As I'm being paid to also work on Qubes as a UX specialist, I'd really love the folks here to understand that we just don't have to make those compromises. Which is not to say the problems aren't "hard." The problems are still hard; they're just solvable, not "Oop, this is a 'security' required thing, so no good ux for our user—and they'll just need to deal." 

Forcing users into experiences where they lose agency, results in compromised opsec practices. There's abundant evidence demonstrating that. There will always be ways to both deliver user agency in a UI (password visibility options), while also managing their expectations (a small 'nudge' line of text to be careful), educating them (a link to an article to learn more about opsec and passwords), and providing them them with a digital experience that puts them in control (a hide/show option). 

Regardless of which screenlocker is ultimately picked—and I'll leave that to the dev experts—please let's all work together to ensure it does not freak-out less technical users, nor encourage poor opsec in all users, for things we could have easily made different choices with.

**Comment by 3hhh** [comment]
> For fancy stuff one could still lock them down and start a dedicated X instance for whatever screensaver the user wishes and then, if that screensaver finished gracefully, unlock the VTs incl. the first X instance again. If the screensaver X instance crashes... blue pill and do it again? At least no windows to restore... ^^

A more detailed implementation sketch for that option:

On VT3 run an application in an endless loop that starts an X server and displays whatever user-defined screensaver.

Implement locking as follows:
1. Switch to VT3 (`chvt 3`).
2. Disable VT switching (`physlock -l` can e.g. do that) & SysRq et al, if enabled.

Implement unlocking to be triggered by a zero exit code of the running screensaver:
1. Re-enable VT switching et al.
2. Switch back to VT1.

Features:
- A failing X or screensaver shouldn't matter as they are just restarted in that endless loop.
- Lock speed should be as fast as VT switching to an already-running screensaver.
- Whatever fancy user-picked screensaver can be used.
- VT1 content cannot be seen during screensaver failures as we're running on VT3.

Some hardening might have to be applied to make it harder to use the time window of an unprotected X server on screensaver restarts (maybe the blue pill approach), but it's probably better than the current fail-open situation.

Any thoughts?

**Comment by marmarek** [comment]
light-locker does similar thing - locks original session with fullscreen window on top (just like xscreensaver) but then starts a new X server switches to it have unlock prompt there (using standard lightdm prompt). It doesn't lock vt switch (which I believe is intentional, to allow other user to log in), but also, whenever you switch to the locked session, it automatically move you to the vt with lightdm prompt (with few seconds delay). Also, my understanding of its architecture is that crashing of password-prompt component (lightdm here) will _not_ give you unlocked session (the unlock requires explicit message "please unlock", not just exiting some process.

So, besides locking vt switch, it does what we need. But that vt switch locking is quite important missing feature. For example, if you try to lock the screen with some (dom0 app) menu open, it will fail at that "fullscreen window" step - just like xscreensaver fails. Then, it will correctly switch to the new X server with the password prompt, but the original X session remains unlocked - if you switch back to it, you get full access. I'd say that's worse that "just" failing to lock, since it gives an impression locking succeeded, while it doesn't.

**Comment by na--** [comment]
For me, light-locker has crashed multiple times in such a way that I've needed to log into a separate VT and kill its process, so that I can "log" back in... I understand that it can be a security issue with the fail-open model, but it can also be a feature that saved me from having to hard-reset the laptop a few times.

**Comment by 3hhh** [comment]
> light-locker does similar thing - locks original session with fullscreen window on top (just like xscreensaver) but then starts a new X server switches to it have unlock prompt there (using standard lightdm prompt). It doesn't lock vt switch (which I believe is intentional, to allow other user to log in), but also, whenever you switch to the locked session, it automatically move you to the vt with lightdm prompt (with few seconds delay). Also, my understanding of its architecture is that crashing of password-prompt component (lightdm here) will _not_ give you unlocked session (the unlock requires explicit message "please unlock", not just exiting some process.
> 
> So, besides locking vt switch, it does what we need. But that vt switch locking is quite important missing feature.

Custom screen lockers are also not supported in that dedicated X session, but indeed it  otherwise does 80% of the job.

Side note:
I recently ran into [this bug](https://bugs.archlinux.org/task/64771), which means that an X restart may crash `xss-lock` and thus prevent screen locking due to user idle times. That can happen with whatever screensaver. I put a note to [the community doc](https://github.com/Qubes-Community/Contents/pull/102).

**Comment by 3hhh** [comment]
I implemented a dom0 PoC for the aforementioned idea [here](https://github.com/3hhh/qubes-issue-1917-poc).

The biggest issue appears to be that Qubes OS fails to manage more than one X server instance. I think that the Qubes GUI daemon attempts to use the last created X server or so (and that's the screen locker X = bad).
It might be worth fixing that as part of a dedicated issue.

In total dom0 applications displaying themselves on the screen locker X server may be an issue.

So currently you'll have to restart your machine after executing the PoC. ^^

Anyway it seems generally feasible.

**Comment by marmarek** [comment]
> The biggest issue appears to be that Qubes OS fails to manage more than one X server instance.

This was the case some time ago, but shouldn't be anymore. It was about `/var/run/qubes/shm.id` file, which was common regardless of `$DISPLAY`, but it isn't the case anymore.

**Comment by 3hhh** [comment]
My tests revealed that starting a second X from `:1` (i.e. from inside the main X) causes the issue to come up. It doesn't come up when one switches to e.g. VT3 manually and then runs e.g. `sudo startx /usr/bin/xterm -- :3 vt3`.

I guess it's related to side effects caused by `startx` resulting in `XOpenDisplay(NULL)` in qubes-guid to fail.
Guid log:
```
No protocol specified
XOpenDisplay: Resource temporarily unavailable
```

**Comment by 3hhh** [comment]
`startx` sets some `xauth`-related environment variables.

Anyway it works flawlessly (i.e. without reboot) from other VTs or with systemd; so essentially it's no issue.

With some good X screensaver that reports its exit status it should be fairly usable.

**Comment by woju** [comment]
FYI there is a recent discussion in linuxmint/cinnamon-screensaver#354, describing Mint's approach to lockers (esp. light-locker). The context is a usual cat-on-a-keyboard bug where children smashed the (virtual) keyboard and the locker crash-unlocked.

**Comment by DemiMarie** [comment]
I consider crash-X-on-screenlocker-exit to be the best option.  Another option is to pass the X connection FD to PID 1 using SCM_RIGHTS; this ensures that the connection will never close until X closes the connection.  (If PID 1 crashes, the kernel panics, so this is not a security issue.)

**Comment by 3hhh** [comment]
> I consider crash-X-on-screenlocker-exit to be the best option.

Killing X will leave the user at the merit of the parent, i.e. the display manager (DM).

What the DM does is highly implementation dependent. In theory it may
1. exit or otherwise fail
2. request authentication and restart X
3. restart X without authentication

Depending on the configuration `lightdm` appears to do 2 or 3 (3 e.g. with autologin enabled or some session restore code). Against 3 you can safeguard by checking whether the screenlocker crashed and start it as part of the X start. From my experience that may sometimes also fail.

To be sure you could kill the DM or even the entire machine with `kill -9 -1`. Actually the latter [is recommended by e.g. google](https://github.com/google/xsecurelock#security-design) for security-conscious users.

Any of these choices result in pretty poor UX though and depend on a lot of code from various authors (X, DM, screenlocker).

VT locking on the other hand will only fail, if the VT locking code is broken. :-)
The PoC also showed that you may still go for the fancy UX stuff with it by only locking VT switches at the cost of having to trust some X and screenlocker code.

Btw I guess most Linux distros didn't go for VT locking so far as it breaks the Linux multi user model. Qubes OS fortunately doesn't have that.

**Comment by iamahuman** [comment]
Bumping this issue as #5460 has been fixed (in testing).

**Comment by kalkin** [comment]
I prefer the ugliness of Xscreensaver to the incompetence of other screen locker developers. For all who think Xscreensaver is ugly: JWZ (the Xscreensaver dev) is waiting for almost two decades, that people who complain about the looks send in patches. This never happened, because it's easier to complain then actually get stuff done.

**Comment by iamahuman** [comment]
@kalkin 

> I prefer the ugliness of Xscreensaver

physlock is a VT-based screen locker (not even a screensaver), which makes it even uglier than XScreenSaver.

> to the incompetence of other screen locker developers.

I assume you're quoting Jamie Zawinski's words. It might be true for most X11-based screensavers, but I don't see how it holds for physlock.

light-locker does use X11, but it also opens a new VT which makes it free from X11 protocol shortcomings (e.g. #2026).

> For all who think Xscreensaver is ugly:

Aesthetics isn't the top priority here--at least compared to security. This is why Qubes is using XScreenSaver now rather than other alternative like gnome-screensaver.

> JWZ (the Xscreensaver dev) is waiting for almost two decades, that people who complain about the looks send in patches. This never happened, because it's easier to complain then actually get stuff done.

**Comment by iamahuman** [comment]
The issue description might need to be updated to reflect the title for minimizing further misunderstandings.

**Comment by ninavizz** [comment]
> Aesthetics isn't the top priority here--at least compared to security. This is why Qubes is using XScreenSaver now rather than other alternatives like gnome-screensaver.


For non-technical users, aesthetics absolutely must be an issue. EG: XScreenSaver with a wall on fire that is rendered c.1992 appears hack-ish and out of place. I know it's developer (Jamie) thinks he's being cute evoking vintage computer graphics, but non-technical users new to security concepts are more likely to find it to be alarming. Yep, I've already reached-out to him about delivering some kind of an update so he didn't have to do any work, and he has zero interest in changing it. FYI, he now owns/manages a nightclub, and has kind of always been dickish about UX stuff from what I've learned; I'm not optimistic that he'll ever update it.

That said, there's also no option to trade-off "incompetence" for "aesthetics." The aesthetics can be _unattractive_, but not alarming. Jamie's stupid choices, make for an alarming screenlocker that'll freak-out the non-technical but high-risk users I'm hoping to get using Qubes.

Are there _any_ alternatives that may be unattractive but are easy to use, competent with security, and not alarming in evoked aesthetics?

**Comment by ninavizz** [comment]
@andrewdavidwong Could we maybe change the name of this Issue to simply "Change default screenlocker from XScreenSaver" ??

**Comment by DemiMarie** [comment]
One idea I had is to pass an X file descriptor to systemd (PID 1).  If PID 1 crashes, the kernel panics.  If it does not, then the connection will stay open forever.

**Comment by unman** [comment]
On Sat, Feb 20, 2021 at 04:45:30PM -0800, Nina Eleanor Alter wrote:
> > Aesthetics isn't the top priority here--at least compared to security. This is why Qubes is using XScreenSaver now rather than other alternatives like gnome-screensaver.
> 
> 
> For non-technical users, aesthetics absolutely must be an issue. EG: XScreenSaver with a wall on fire that is rendered c.1992 appears hack-ish and out of place. I know it's developer (Jamie) thinks he's being cute evoking vintage computer graphics, but non-technical users new to security concepts are more likely to find it to be alarming. Yep, I've already reached-out to him about delivering some kind of an update so he didn't have to do any work, and he has zero interest in changing it. FYI, he now owns/manages a nightclub, and has kind of always been dickish about UX stuff from what I've learned; I'm not optimistic that he'll ever update it.
> 
> That said, there's also no option to trade-off "incompetence" for "aesthetics." The aesthetics can be _unattractive_, but not alarming. Jamie's stupid choices, make for an alarming screenlocker that'll freak-out the non-technical but high-risk users I'm hoping to get using Qubes.
> 
> Are there _any_ alternatives that may be unattractive but are easy to use, competent with security, and not alarming in evoked aesthetics?
> 

The problem is that XScreenSaver is the best of the available options -
that's why it has lasted so long.
I'm really not aware of any one who has *ever* been alarmed by the wall
of fire, or has been freaked out. I *do* know people who have been
thrown by some of the hacks.
But if this is considered to be a real issue then why not just fork and
replace the offending wall with some other (equally basic) image? The
framework is great, and we should keep it. (I've started testing v6
which is significant rewrite, but the wall is still there.)

**Comment by andrewdavidwong** [comment]
> @andrewdavidwong Could we maybe change the name of this Issue to simply "Change default screenlocker from XScreenSaver" ??

Yep, good idea. (I thought I had already done so, but I must've been thinking about a different issue.)

**Comment by andrewdavidwong** [comment]
> But if this is considered to be a real issue then why not just fork and replace the offending wall with some other (equally basic) image?

I suppose the obvious choice for a replacement image would be the Qubes logo (if that's permitted by licensing and such).

**Comment by marmarek** [comment]
> One idea I had is to pass an X file descriptor to systemd (PID 1). If PID 1 crashes, the kernel panics. If it does not, then the connection will stay open forever.

No this won't solve anything and is a bad idea in general. First, "lets crash the whole system" should be the last resort if there is really not other way to recover. In case of multi-user system (not Qubes right now, but maybe in the future), ability for a misbehaving application to crash the whole system is actually considered security issue. Lets not add things that may bite us.
Furthermore, crashing X server is not a problem in this issue. The X server staying alive and accessible when the screen locker crashes (or fails to start) is. Killing X server if the screen locker fails to start may be an option, but only because it does not kill user applications ("just" makes it a bit inconvenient, as window locations are lost).

**Comment by 3hhh** [comment]
This issue might be something for GSOC.

**Comment by ninavizz** [comment]
If replacing just the flame illustration would be possible (on both the fail screen and the start screen) the gnome "computer" or "screen" icons would work splendidly; or the Qubes logo, as Andrew suggests. The prior I think would be less controversial—but I'm increasingly leaning towards Andrew's suggestion.

The sense I got, is that Jamie could really care less wrt copyright stuff. 

@unman My own initial first impression of the screensaver, was that it was a hack; mostly because ALL of its UI felt so dated, and thus that was the only explanation. Such suspicions are also a known pattern from UX research of vulnerable users with security software.

**Comment by DemiMarie** [comment]
> > One idea I had is to pass an X file descriptor to systemd (PID 1). If PID 1 crashes, the kernel panics. If it does not, then the connection will stay open forever.
> 
> No this won't solve anything and is a bad idea in general. First, "lets crash the whole system" should be the last resort if there is really not other way to recover. In case of multi-user system (not Qubes right now, but maybe in the future), ability for a misbehaving application to crash the whole system is actually considered security issue. Lets not add things that may bite us.

That was poorly phrased.  On Linux, PID 1 in the initial PID namespace is the one process that is guaranteed to stay up as long as the system is running, since the kernel panics if it ever exits.  Furthermore, we use systemd for PID 1, which provides an API for storing arbitrary file descriptors in it, and a connection to the X server is just a file descriptor!

Therefore, we can pass the file descriptor for the X connection to PID 1.  Since PID 1 will never exit, this file descriptor will never be closed, so X will not realize that the screensaver has exited.  If the screensaver exits, something has *already* gone seriously wrong, and I consider killing the X server to be okay in this case.

> Furthermore, crashing X server is not a problem in this issue. The X server staying alive and accessible when the screen locker crashes (or fails to start) is. Killing X server if the screen locker fails to start may be an option, but only because it does not kill user applications ("just" makes it a bit inconvenient, as window locations are lost).

This doesn’t help the “screen locker doesn’t start” case, but it might solve the “screen locker crashes” case.  I doubt the “screen locker doesn’t start” case will be solved until we transition to Wayland.

**Comment by marmarek** [comment]
> I doubt the “screen locker doesn’t start” case will be solved until we transition to Wayland.

The main (the only?) reason in something holding a server (or keyboard, or pointer) grab, which is the case when some dom0 application opens a menu (at least GTK does grab keyboard then). We (obviously?) don't allow VM applications doing any kind of grab in dom0's X server, but applications running directly in dom0 (window manager related) are free to do so.

This failure mode can be solved by doing VT switch, regardless of what you are switching to. Physlock does that by switching to a text console and preventing switching VT until unlocked. Light-locker does that by starting another X server instance, but sadly it doesn't have an option to prevent switching back (to unlocked X session in such failure case).

**Comment by ninavizz** [comment]
Uncertain where the "status" on this Issue is... but I did go ahead and create #6425, on the chance that leaving XScreenSaver as QubesOS' screen-locker is the decision for now. FYI.

**Comment by besendorf** [comment]
XScreensaver has been refactored quite significantly with the purpose to reduce the code in the critical section and make it more secure. See jwz's blogpost here: https://www.jwz.org/blog/2021/02/xscreensaver-version-6-beta-1/
Maybe you should reevaluate changing the screensaver based on this. I can't tell because I don't know enough about the X windows system.

**Comment by unman** [comment]
On Sat, Feb 27, 2021 at 06:50:01AM -0800, besendorf wrote:
> XScreensaver has been refactored quite significantly with the purpose to reduce the code in the critical section and make it more secure. See jwz's blogpost here: https://www.jwz.org/blog/2021/02/xscreensaver-version-6-beta-1/
> Maybe you should reevaluate changing the screensaver based on this. I can't tell because I don't know enough about the X windows system.
> 

Indeed, and I specifically referred to this new version, in the linked
issue - #6425
As far as I can see the main argument against XScreenLocker has, in the
end, come down to UX - although there are some other issues, it remains
the best locker out there imo.
I'm using the test release now, and we are waiting on a decision whether
to rebrand and release qscreenlocker with 4.1.

**Comment by 3hhh** [comment]
> As far as I can see the main argument against XScreenLocker has, in the
end, come down to UX [...]

That's incorrect. Initially the issue was created for security reasons (`physlock` is not better wrt UX). @ninavizz just made her point that UX is also an issue with `xscreensaver` later on.

As pointed out earlier, all screen lockers based on X will have focus issues, X crashing issues, ...
Making whatever other security relevant code smaller won't help there.

**Comment by unman** [comment]
Simply not true: the issue was *originally* raised because the lock
screen looked "wacky-different, also ugly", and was entirely a UX issue.
Discussion of the security issues followed sometime later, much of it
confused or misplaced.

I continue to think that Xscreensaver is the best option available. If
the UX is held to be so bad/confusing that we should rebrand, we can do
that simply enough.

**Comment by iamahuman** [comment]
How about using light-locker while letting xscreensaver grab the original X display?

**Comment by Edan-Redmind** [comment]
@ninavizz 
Sadly, I didn't have time to read all of the excellent points raised here.
Just wanted to point out that current lock screen has no indication of what Language input the keyboard is set, so people who use two or more languages get multiple (frustrating) “Authentication failed!” messages and have to blindly change language inputs until they get the right one.
cheers.

**Comment by ninavizz** [comment]
@Edan-Redmind I just made a comment on an issue here for a forked version of XScreenSaver, about language I'd like to updated for _English_ usability in that... and was wondering about how this worked in different languages. Not sure how this can be addressed when Qubes is localized... I don't even know if XScreenSaver has been looked at, for localization. @andrewdavidwong? 

Clearly this will be a thing that needs to get resolved at some point. Thank you for flagging, knowing what issues exist today is very helpful!

**Comment by Edan-Redmind** [comment]
@ninavizz Thank you. I would like to point out that I wasn't talking about localized version of Qubes, Everything here is in english, just with a second input-language on the keyboard. So this is relevant to a fair everyday-use (email?), in every country that use non-english as it's native language.

**Comment by na--** [comment]
As I wrote in https://github.com/QubesOS/qubes-issues/issues/1917#issuecomment-736806749, I'm having the same issue with multiple [keyboard layouts](https://www.qubes-os.org/doc/installation-guide/#localization) (not with English error messages...), with an extra bonus for XScreensaver not showing the caps lock status and my laptop not having a visual indicator for it.

To me the logo question seems like a minor bikesheddy concern that, when you draw the line, wouldn't solve any of the actual UX or security problems anyone is having with QubesOS, while taking up time (all of these discussions) and introducing more maintenance burdens (maintaining a separate xscreensaver package and/or fork). Certainly worth fixing if everything else was perfect, but that's far from the case...

**Comment by andrewdavidwong** [comment]
> @Edan-Redmind I just made a comment on an issue here for a forked version of XScreenSaver, about language I'd like to updated for _English_ usability in that... and was wondering about how this worked in different languages. Not sure how this can be addressed when Qubes is localized... I don't even know if XScreenSaver has been looked at, for localization. @andrewdavidwong?
> 

This would be handled by the localization team. This project has not taken any action on localizing XScreenSaver, since it is not our software, and there is no QScreenSaver yet.

**Comment by unman** [comment]
On Mon, Mar 15, 2021 at 02:53:11PM -0700, Andrew David Wong wrote:
> > @Edan-Redmind I just made a comment on an issue here for a forked version of XScreenSaver, about language I'd like to updated for _English_ usability in that... and was wondering about how this worked in different languages. Not sure how this can be addressed when Qubes is localized... I don't even know if XScreenSaver has been looked at, for localization. @andrewdavidwong?
> > 
> 
> This would be handled by the localization team. This project has not taken any action on localizing XScreenSaver, since it is not our software, and there is no QScreenSaver yet.
> 
> 

XScreenSaver is (I think) fully localized as is, so I think this need
not be handled by our localization team.

**Comment by ninavizz** [comment]
I see this issue has a 4.1 milestone on it. I'd rather not see the 4.1 release exceed what it's promised (which doesn't include this) as I suspect users are really looking forward to all else it promised. If 4.2 is really UI centered, this would fit in really nicely, there, with the updated App Menu, Q Manager (not documented in GH yet) and any other UX or UI changes.

**Comment by mfc** [comment]
hi folks, it doesn't seem like we are following XFCE developments enough. 

Qubes release 4.0 has XFCE 4.12.

the very next version of XFCE 4.14 apparently fixes the UX issue with updated xscreensaver, see the [XFCE 4.14 tour screen](https://xfce.org/about/tour414), specifically "The new lock screen, sharing styles with LightDM GTK+ Greeter for a consistent experience.".

if that's the case, then to solve this and https://github.com/QubesOS/qubes-issues/issues/6425 we can just ship a newer XFCE + xscreensaver and have the user icon by default be Qubes logo, no need to tweak xscreensaver.

**Comment by marmarek** [comment]
I'm a bit lost about relation between xfce4-screensaver, light-locker and xscreensaver here. We do have XFCE 4.14 in Qubes 4.1, so I can try this out (although not at this moment). Xscreensaver used there directly looks identical as in Qubes 4.0, but maybe xfce ships some wrapper?
Anyway, the graphical aspect is only one thing. The more important one is actually locking the screen and then being able to unlock it. I can apply my usual testing routine (which also includes "cat" and "keyboard" ;) ) on xfce4-screensaver and see how it stands. Both GNOME and KDE defaults fail this test.

**Comment by unman** [comment]
On Sat, Mar 27, 2021 at 07:09:59AM -0700, Marek Marczykowski-G??recki wrote:
> I'm a bit lost about relation between xfce4-screensaver, light-locker and xscreensaver here. We do have XFCE 4.14 in Qubes 4.1, so I can try this out (although not at this moment). Xscreensaver used there directly looks identical as in Qubes 4.0, but maybe xfce ships some wrapper?
> Anyway, the graphical aspect is only one thing. The more important one is actually locking the screen and then being able to unlock it. I can apply my usual testing routine (which also includes "cat" and "keyboard" ;) ) on xfce4-screensaver and see how it stands. Both GNOME and KDE defaults fail this test.

As I pointed out over at #6425, this isn't a version of Xscreensaver -
it's a port of GNOME screensaver. We don't want that.

**Comment by 3hhh** [comment]
Since apparently everyone gave up on attempting to fix the underlying X issues, you could pretty much consider _any_ X-based screensaver that fulfils the following requirements:

- succeeds at basic locking/unlocking tests (cat test)
- provides at least similar security standards as `xscreensaver`
- "doesn't look like malware" (UX)
- has some i18n features (UTF-8 support and language switch should suffice)
- ideally is in Fedora dom0 repos so that it doesn't require extra maintenance
- should have active maintenance to get upstream bug fixes

I'd recommend doing a comparison based on such pre-defined criteria to finally come to some conclusion.

E.g. IIRC both [xsecurelock](https://github.com/google/xsecurelock) and [light-locker](https://github.com/the-cavalry/light-locker) should be available in 4.1 dom0 repos; so it should be rather straightforward to test them.

**Comment by unman** [comment]
On Sun, Mar 28, 2021 at 03:44:08AM -0700, 3hhh wrote:
> Since apparently everyone gave up on attempting to fix the underlying X issues, you could pretty much consider _any_ X-based screensaver that fulfils the following requirements:
> 
> - succeeds at basic locking/unlocking tests (cat test)
> - provides at least similar security standards as `xscreensaver`
> - "doesn't look like malware" (UX)
> - has some i18n features (UTF-8 support and language switch should suffice)
> - ideally is in Fedora dom0 repos so that it doesn't require extra maintenance
> - should have active maintenance to get upstream bug fixes
> 
> I'd recommend doing a comparison based on such pre-defined criteria to finally come to some conclusion.
> 
> E.g. IIRC both [xsecurelock](https://github.com/google/xsecurelock) and [light-locker](https://github.com/the-cavalry/light-locker) should be available in 4.1 dom0 repos; so it should be rather straightforward to test them.
> 

There is already a proposal on the table to use Xscreensaver,
patched so it "doesn't look like malware".
There is no need for any solution to be in the Fedora repos - we already
ship a patched Xscreensaver to remove the "offending" update reminder.
I'm not clear why  any solution would need a "language switch" if by
that you mean the capability to change languages *in the screenlocker*
on the fly before inputting a passphrase. Is this common? Can you point
me to an example?

**Comment by Edan-Redmind** [comment]
@3hhh
I use english-only passwords but have secondary language installed. having some indicator for current language will help avoid "wrong pw msgs" when i don't recall what was the last input language i used when i left the workstation.
________________________________

**Comment by unman** [comment]
> @3hhh I use english-only passwords but have secondary language installed. having some indicator for current language will help avoid "wrong pw msgs" when i don't recall what was the last input language i used when i left the workstation.

I understand the problem.
My question was about the proposed:
>  has some i18n features (UTF-8 support and language switch should suffice)

I struggle to understand what "language switch" means here - an example from another locker would help.

**Comment by ninavizz** [comment]
^^ +1 to the question around language w/ lockers

**Comment by na--** [comment]
When you want to be able to type text in multiple languages, you do that by having multiple keyboard layouts on your system: https://docs.xfce.org/xfce/xfce4-settings/keyboard#layout

Keyboard layouts allow you to switch and be able to use the same keys on the same keyboard to type in different languages. You usually shift the layout by some shortcut (e.g. Ctrl+Shift), and typing "test" becomes "тест" or whatever. You also usually have a widget in the system tray that shows the currently active keyboard layout (as a flag or a language code or whatever you configure). QubesOS also has support for propagating the currently active keyboard layout from dom0 to VMs and that works reasonably ok.

Unfortunately, XScreenSaver is the worst possible screen locker in terms of password input UX, because it doesn't have any indication what keyboard layout you're currently using when typing your password. So if your password is `TestPassword`, you might be typing `TestPassword` or `ТестПассворд`, or even something else, if you have more than 2 keyboard layouts enabled. And you have to guess blindly - when it complains that your password is wrong, you don't know if you mistyped it or if you were in the wrong keyboard layout state.

As I explained above (https://github.com/QubesOS/qubes-issues/issues/1917#issuecomment-736806749), my situation is even worse, because my laptop doesn't have visual indicators for the capslock or numlock state, and re-uses part of the keyboard as the numlock :man_facepalming: . So, when typing my `TestPassword` password, I can be in one of 8 possible states and type any one of these: `TestPassword`, `tESTpASSWORD` (caps), `Test/assw6rd` (num), `tEST/ASSW6RD` (caps+num), `ТестПассворд` (bg), `тЕСТпАССВОРД` (bg+caps), `Тест/ассв6рд` (bg+num), `tEST/АССВ6РД` (bg+caps+num). And, because XScreenSaver doesn't have any indication of the current keyboard state or layout, you have to blindly guess and toggle things on and off until you hit the correct one... Also, imagine what happens if you were in the correct state, by some miracle, but you actually mistype your password, because you are security-conscious person and have a reasonably complex one...

Pretty much all of the other screen lockers have a visual indication of at least the currently active keyboard layout and caps lock state. It makes a huge difference.

**Comment by na--** [comment]
Some examples from random locker screenshots found on the internet:
- see `en_US` in the upper right corner: https://xubuntu.org/wp-content/uploads/2014/03/xubuntu_lock_light-locker.png
- see `en` in the upper right corner again: https://vitux.com/wp-content/uploads/2018/08/word-image-94-768x432.png
- not sure about kde, but at least you seem to have the eye button that allows you to see what you've typed, so you can easily type `test` and see what state you're in without compromising your actual password: `https://phabricator.kde.org/file/data/utcmshse2pn5j4vgyijj/PHID-FILE-nnprs6hnzbybuk6gr52q/KScreenLocker.png`

**Comment by marmarek** [comment]
In short: it isn't really about language switch, but about keyboard layout switch.

**Comment by unman** [comment]
On Mon, Mar 29, 2021 at 04:04:03AM -0700, Marek Marczykowski-G??recki wrote:
> In short: it isn't really about language switch, but about keyboard layout switch.
> 

Or display of current keyboard layout

**Comment by cfcs** [comment]
> On Mon, Mar 29, 2021 at 04:04:03AM -0700, Marek Marczykowski-G??recki wrote: In short: it isn't really about language switch, but about keyboard layout switch.
> Or display of current keyboard layout

Most users will not know what to do with that information if the keyboard layout is not the one they expect (ie the one printed on the keyboard).

**Comment by ninavizz** [comment]
Ok, so maybe we should create a new issue for _QScreenSaver_ and track these feature needs, there? Only suggesting so important things for it, get tracked there.

**Comment by unman** [comment]
On Fri, Apr 02, 2021 at 12:51:00PM -0700, cfcs wrote:
> > On Mon, Mar 29, 2021 at 04:04:03AM -0700, Marek Marczykowski-G??recki wrote: In short: it isn't really about language switch, but about keyboard layout switch.
> > Or display of current keyboard layout
> 
> Most users will not know what to do with that information if the keyboard layout is not the one they expect (ie the one printed on the keyboard).
> 

I'm trying (and failing ) to find a case where a user has triggered a
screensaver (of any sort) on a keyboard which does not match the
language they are using **without** already being able to switch layout.
Can you help me understand how that happens?

**Comment by cfcs** [comment]
@unman: So I've had to help Qubes users who were unable to unlock the screen where the solution was to type their password with `us` layout. I'm not sure how they ended up in that situation - if they pressed some magic combination to change the dom0 layout, or if they've perhaps just assigned local layouts to all their appvms and not noticed that dom0 was not using the local one (or didn't expect it). All I can tell you is that it's a real problem for some people, and it's not obvious that they suddenly need to type `-` to get a `/` instead of `shift + 7`, so even if they can see that `us` keyboard is used, it won't help them get back to work.

@na-- If you need a quick introduction to how typing works, regardless of keyboard layout, I'll help you: You press the buttons in front of you and text appears. It's definitely not for everybody, but it just might help make you come across less imbecile.

**Comment by unman** [comment]
On Sat, Apr 03, 2021 at 07:03:09AM -0700, cfcs wrote:
> @unman: So I've had to help Qubes users who were unable to unlock the screen where the solution was to type their password with `us` layout. I'm not sure how they ended up in that situation - if they pressed some magic combination to change the dom0 layout, or if they've perhaps just assigned local layouts to all their appvms and not noticed that dom0 was not using the local one (or didn't expect it). All I can tell you is that it's a real problem for some people, and it's not obvious that they suddenly need to type `-` to get a `/` instead of `shift + 7`, so even if they can see that `us` keyboard is used, it won't help them get back to work.
> 

I'm still struggling to understand how they *got* to that state - how
did they log in in the first place? It doesnt look like a screensaver
problem to me.

> @na-- If you need a quick introduction to how typing works, regardless of keyboard layout, I'll help you: You press the buttons in front of you and text appears. It's definitely not for everybody, but it just might help make you come across less imbecile.
> 

I dont know who this comment is aimed at, but it's out of place here.

**Comment by Edan-Redmind** [comment]
@cfcs
very easy. you start working, write an email in a non-englisg layout, then answer the phone and when you're done the screensaver is on and you need to unlock. now, was your last text was the email body (in a non-english layout) or was it the email address in english...? ;)
just an example that happens to me.
________________________________

**Comment by na--** [comment]
> @na-- If you need a quick introduction to how typing works, regardless of keyboard layout, I'll help you: You press the buttons in front of you and text appears. It's definitely not for everybody, but it just might help make you come across less imbecile.

@cfcs, I was about to comment on your [previous remarks](https://github.com/QubesOS/qubes-issues/issues/1917#issuecomment-812684942) earlier today, but deleted my reply just before posting it and left a simple :-1: instead, since I decided to give you the benefit of the doubt that you're not a condescending asshole. Sorry, my mistake... :man_shrugging: 

@unman, usually I get to the "unknown keyboard layout in XScreensaver" state exactly how @Edan-Redmind described. Not by locking my screen explicitly, but by the screen locker engaging automatically after some time has passed without any input. I guess one "solution" would be to edit `xflock4` to always reset my keyboard layout to `us` before it actually locks the screen... :sweat_smile:

**Comment by DemiMarie** [comment]
This is fixed-by-wayland because Wayland actually provides decent ways to implement screenlockers, such that if the screenlocker crashes then either the session is terminated or a new screenlocker is respawned.

**Comment by andrewdavidwong** [comment]
Based on https://github.com/QubesOS/qubes-issues/issues/9317#issuecomment-2964357263, it sounds like this issue might be closed by https://github.com/marmarek/qubes-dist-upgrade/commit/03cba087127c8d335b5f13fa79ea322692b390e5.

**Comment by marmarek** [comment]
R4.3 uses xfce4-screensaver in Xfce by default now.
